ruby/ql/lib/codeql/ruby/ast/Pattern.qll

```diff
@@ -97,3 +97,302 @@ class TuplePattern extends Pattern, TTuplePattern {
 
   override AstNode getAChild(string pred) { pred = "getElement" and result = this.getElement(_) }
 }
+
+private class TPatternNode =
+  TArrayPattern or TFindPattern or THashPattern or TAlternativePattern or TAsPattern or
+      TVariableReferencePattern;
+
+private class TPattern =
+  TPatternNode or TLiteral or TLambda or TConstantAccess or TLocalVariableAccess or
+      TUnaryArithmeticOperation;
+
+/**
+ * A pattern used in a `case-in` expression. For example
+ * ```rb
+ * case expr
+ *   in [ x ] then ...
+ *   in Point(a:, b:) then ...
+ *   in Integer => x then ...
+ * end
+ * ```
+ */
+class CasePattern extends AstNode, TPattern {
+  CasePattern() { casePattern(toGenerated(this)) }
+}
+
+/**
+ * An array pattern, for example:
+ * ```rb
+ * in []
+ * in ["first", Integer => x, "last"]
+ * in ["a", Integer => x, *]
+ * in ["a", Integer => x, ]
+ * in [1, 2, *x, 7, 8]
+ * in [*init, 7, 8]
+ * in List["a", Integer => x, *tail]
+ * ```
+ */
+class ArrayPattern extends CasePattern, TArrayPattern {
+  private Ruby::ArrayPattern g;
+
+  ArrayPattern() { this = TArrayPattern(g) }
+
+  /** Gets the class this pattern matches objects against, if any. */
+  ConstantReadAccess getClass() { toGenerated(result) = g.getClass() }
+
+  /**
+   * Get the `n`th element of this list pattern's prefix, i.e. the elements `1, 2, 3`
+   * in the following examples:
+   * ```
+   * in [ 1, 2, 3 ]
+   * in [ 1, 2, 3, ]
+   * in [ 1, 2, 3, *, 4 , 5]
+   * in [ 1, 2, 3, *more]
+   * ```
+   */
+  CasePattern getPrefixElement(int n) {
+    toGenerated(result) = g.getChild(n) and
+    (
+      n < this.restIndex()
+      or
+      not exists(restIndex())
+    )
+  }
+
+  /**
+   * Get the `n`th element of this list pattern's suffix, i.e. the elements `4, 5`
+   * in the following examples:
+   * ```
+   * in [ *, 4, 5 ]
+   * in [ 1, 2, 3, *middle, 4 , 5]
+   * ```
+   */
+  CasePattern getSuffixElement(int n) { toGenerated(result) = g.getChild(n + this.restIndex() + 1) }
+
+  /** Get the variable of the rest token, if any, i.e. `name` in `*name`. */
```

Comment:
I think the example should include the full array pattern, e.g. `in [ 1, 2, 3, *middle, 4 , 5]`